### PR TITLE
Interface concentration discontinuities (fenicsx)

### PR DESCRIPTION
**Pull request by RemDelaporteMathurin**
## Proposed changes

A massive thank you to @jorgensd who was **instrumental** in getting this functionality in FESTIM 🥇

This PR is a first (big) step towards interface discontinuities in the `fenicsx` branch (#719).

(requires #764) 

This new approach to conservation of chemical potential relies on the mixed domain approach implemented in dolfinx 0.8.0 and based off a demo available [here](https://gist.github.com/jorgensd/8a5c32f491195e838f5863ca88b27bce#file-problem-py)

I created a new problem class `HTransportProblemDiscontinuous` - which is not yet complete - to avoid breaking existing functionality. We will complete this class by integrating transient, post-processing, ICs, etc. Eventually this whole class will be merged with `HydrogenTransportProblem`.

In the current interface, users need to provide correspondence between surfaces and volumes as well as listing interfaces and which volumes they are connected to:

```python
list_of_interfaces = [F.Interface(6, (left_domain, middle_domain))]
my_model.subdomains = [left_domain, right_domain, left_surface, right_surface]
```

I believe this can be streamlined by FESTIM under the hood but I'll keep that for a future PR.

## Proposed user interface:

```python
import numpy as np

import festim as F

my_model = F.HTransportProblemDiscontinuous()

interface_1 = 0.2
interface_2 = 0.8

vertices = np.concatenate(
    [
        np.linspace(0, interface_1, num=100),
        np.linspace(interface_1, interface_2, num=100),
        np.linspace(interface_2, 1, num=100),
    ]
)

my_model.mesh = F.Mesh1D(vertices)

material_left = F.Material(D_0=2.0, E_D=0.1, K_S_0=2.0, E_K_S=0)
material_mid = F.Material(D_0=2.0, E_D=0.1, K_S_0=4.0, E_K_S=0)
material_right = F.Material(D_0=2.0, E_D=0.1, K_S_0=6.0, E_K_S=0)

left_domain = F.VolumeSubdomain1D(3, borders=[0, interface_1], material=material_left)
middle_domain = F.VolumeSubdomain1D(
    4, borders=[interface_1, interface_2], material=material_mid
)
right_domain = F.VolumeSubdomain1D(5, borders=[interface_2, 1], material=material_right)

left_surface = F.SurfaceSubdomain1D(id=1, x=vertices[0])
right_surface = F.SurfaceSubdomain1D(id=2, x=vertices[-1])

# the ids here are arbitrary in 1D, you can put anything as long as it's not the same as the surfaces
my_model.interfaces = [
    F.Interface(6, (left_domain, middle_domain)),
    F.Interface(7, (middle_domain, right_domain)),
]
my_model.surface_to_volume = {right_surface: right_domain, left_surface: left_domain}

my_model.subdomains = [
    left_domain,
    middle_domain,
    right_domain,
    left_surface,
    right_surface,
]

H = F.Species("H", mobile=True)
trapped_H = F.Species("H_trapped", mobile=False)
empty_trap = F.ImplicitSpecies(n=0.5, others=[trapped_H])

my_model.species = [H, trapped_H]

for species in [H, trapped_H]:
    species.subdomains = [left_domain, middle_domain, right_domain]


my_model.reactions = [
    F.Reaction(
        reactant=[H, empty_trap],
        product=[trapped_H],
        k_0=2,
        E_k=0,
        p_0=0.1,
        E_p=0,
        volume=domain,
    )
    for domain in [left_domain, middle_domain, right_domain]
]

my_model.boundary_conditions = [
    F.DirichletBC(left_surface, value=0.05, species=H),
    F.DirichletBC(right_surface, value=0.2, species=H),
]


my_model.temperature = lambda x: 300 + 100 * x[0]

my_model.settings = F.Settings(atol=1e-10, rtol=1e-10, transient=False)

my_model.exports = [
    F.VTXSpeciesExport(
        f"u_{field.name}_{subdomain.id}.bp", field=field, subdomain=subdomain
    )
    for subdomain in my_model.volume_subdomains
    for field in [H, trapped_H]
]

my_model.initialise()
my_model.run()
```

image

## Types of changes

What types of changes does your code introduce to FESTIM?

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Code refactoring
- [ ] Documentation Update (if none of the other choices apply)
- [ ] New tests

## Checklist

- [x] Black formatted
- [x] Unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

**Comment by RemDelaporteMathurin** [comment]
We have noticed crashed in 1D simulations with coarse meshes. This is [a fix for it](https://github.com/FEniCS/dolfinx/pull/3361) (needs testing and review) and the working example is [here](https://gist.github.com/jorgensd/29ca28ce55c4da3dba0e7031993f997e) (works with this PR).

**Comment by RemDelaporteMathurin** [comment]
We had a discussion with @jorgensd @jhdark and @JonathanGSDUFOUR and noticed weird things happening with the performance so we need to investigate. Otherwise I'll carry on writing tests for this

File: festim/helpers_discontinuity.py [hunk 1]
```diff
@@ -0,0 +1,223 @@
+from mpi4py import MPI
+import dolfinx
+import dolfinx.fem.petsc
+import ufl
+import numpy as np
+from petsc4py import PETSc
+import basix
+
+
+class NewtonSolver:
```
Comment (RemDelaporteMathurin) [inline festim/helpers_discontinuity.py:10]: Eventually we will replace this by the solver provided by https://github.com/scientificcomputing/scifem

File: festim/mesh/mesh.py [hunk 2]
```diff
@@ -86,10 +88,29 @@ def define_meshtags(self, surface_subdomains, volume_subdomains):
...
+                self.mesh.topology,
+                volume_meshtags.find(domain_2.id),
+                self.vdim,
+                self.fdim,
+            )
+            interface_entities = np.intersect1d(all_1_facets, all_2_facets)
+            tags_facets[interface_entities] = interface.id
+
+        facet_meshtags = meshtags(self.mesh, self.fdim, mesh_facet_indices, tags_facets)
+
```
Comment (RemDelaporteMathurin) [inline festim/mesh/mesh.py:115]: TODO: This is not covered because the tests only cover 1D meshes.

Comment (RemDelaporteMathurin) [inline festim/mesh/mesh.py:115]: The reason why these lines are not covered is because the only multi-material case has meshtags provided by the user so `define_meshtags` isn't called at all.

We would need a 1D test case

File: festim/hydrogen_transport_problem.py [hunk 3]
```diff
@@ -711,3 +713,450 @@ def post_processing(self):
...
+            )
+        subdomain.u = u
+        subdomain.u_n = u_n
+
+    def create_subdomain_formulation(self, subdomain: F.VolumeSubdomain):
+        form = 0
+        # add diffusion and time derivative for each species
+        for spe in self.species:
+            if subdomain not in spe.subdomains:
+                continue
```
Comment (RemDelaporteMathurin) [inline festim/hydrogen_transport_problem.py:892]: To test this we need to have a species not existing in every subdomain

File: festim/hydrogen_transport_problem.py [hunk 4]
```diff
@@ -711,3 +713,450 @@ def post_processing(self):
...
+            jac = []
+            form = subdomain1.F
+            for subdomain2 in self.volume_subdomains:
+                jac.append(
+                    dolfinx.fem.form(
+                        ufl.derivative(form, subdomain2.u), entity_maps=entity_maps
+                    )
+                )
+            J.append(jac)
+            forms.append(dolfinx.fem.form(subdomain1.F, entity_maps=entity_maps))
```
Comment (RemDelaporteMathurin) [inline festim/hydrogen_transport_problem.py:1065]: Thanks to https://github.com/FEniCS/dolfinx/pull/3450 we should be able to simplify this bit

File: discontinuity_generic.py [hunk 5]
```diff
@@ -0,0 +1,644 @@
...
+
+
+def transfer_meshtags_to_submesh(
+    mesh, entity_tag, submesh, sub_vertex_to_parent, sub_cell_to_parent
+):
+    """
+    Transfer a meshtag from a parent mesh to a sub-mesh.
+    """
+
+    tdim = mesh.topology.dim
```
Comment (jorgensd) [inline discontinuity_generic.py:175]:
```suggestion
   # Invert map from sub-mesh to parent-mesh to parent-mesh to sub-mesh
    tdim = mesh.topology.dim
```

File: discontinuity_generic.py [hunk 6]
```diff
@@ -0,0 +1,644 @@
...
+    cell_imap = mesh.topology.index_map(tdim)
+    num_cells = cell_imap.size_local + cell_imap.num_ghosts
+    mesh_to_submesh = np.full(num_cells, -1)
+    mesh_to_submesh[sub_cell_to_parent] = np.arange(
+        len(sub_cell_to_parent), dtype=np.int32
+    )
+    sub_vertex_to_parent = np.asarray(sub_vertex_to_parent)
```
Comment (jorgensd) [inline discontinuity_generic.py:182]:
```suggestion
```
I don't think we need this anymore after: https://github.com/FEniCS/dolfinx/pull/3112/

File: discontinuity_generic.py [hunk 7]
```diff
@@ -0,0 +1,644 @@
...
+
+    submesh.topology.create_connectivity(entity_tag.dim, 0)
```
Comment (jorgensd) [inline discontinuity_generic.py:184]: # Compute and access various entity to vertex and vertex to entity connectivity for parent and sub-mesh.

File: discontinuity_generic.py [hunk 8]
```diff
@@ -0,0 +1,644 @@
...
+
+    tdim = mesh.topology.dim
+    cell_imap = mesh.topology.index_map(tdim)
+    num_cells = cell_imap.size_local + cell_imap.num_ghosts
+    mesh_to_submesh = np.full(num_cells, -1)
+    mesh_to_submesh[sub_cell_to_parent] = np.arange(
+        len(sub_cell_to_parent), dtype=np.int32
+    )
+    sub_vertex_to_parent = np.asarray(sub_vertex_to_parent)
+
```
Comment (jorgensd) [inline discontinuity_generic.py:183]:
```suggestion
# Compute and access various entity to vertex and vertex to entity connectivity for parent and sub-mesh.
```

File: festim/helpers_discontinuity.py [hunk 9]
```diff
@@ -0,0 +1,223 @@
+from mpi4py import MPI
+import dolfinx
+import dolfinx.fem.petsc
+import ufl
+import numpy as np
+from petsc4py import PETSc
+import basix
```
[the remark this answers is not in the document]
Comment (jorgensd) [reply festim/helpers_discontinuity.py:7]: Switch import order to ensure that PETSc/MPI and DOLFINx doesn't argue about MPI initialization.

File: festim/hydrogen_transport_problem.py [hunk 10]
```diff
@@ -711,3 +713,506 @@ def post_processing(self):
...
+            self.define_function_spaces(subdomain)
+            ct_r = dolfinx.mesh.meshtags(
+                self.mesh.mesh,
+                self.mesh.mesh.topology.dim,
+                subdomain.submesh_to_mesh,
+                np.full_like(subdomain.submesh_to_mesh, 1, dtype=np.int32),
+            )
+            subdomain.dx = ufl.Measure(
+                "dx", domain=self.mesh.mesh, subdomain_data=ct_r, subdomain_id=1
+            )
```
Comment (jorgensd) [inline festim/hydrogen_transport_problem.py:825]:
```suggestion
            subdomain_dx = self.parent_dx(subdomain_marker)
```
If we define a `parent_dx` in the initialization of the TransportProblem.
We sit down and fix this later today.

File: festim/hydrogen_transport_problem.py [hunk 11]
```diff
@@ -711,3 +713,506 @@ def post_processing(self):
...
+        cr = ufl.Circumradius(mesh)
+
+        entity_maps = {
+            sd.submesh: sd.parent_to_submesh for sd in self.volume_subdomains
+        }
+        for interface in self.interfaces:
+            gamma = interface.penalty_term
+
+            subdomain_1, subdomain_2 = interface.subdomains
+            b_res, t_res = interface.restriction
```
Comment (RemDelaporteMathurin) [inline festim/hydrogen_transport_problem.py:1053]: t and b here refer to bottom and top from a previous exmaple, should change

File: festim/hydrogen_transport_problem.py [hunk 12]
```diff
@@ -711,3 +713,506 @@ def post_processing(self):
...
+        for subdomain1 in self.volume_subdomains:
+            jac = []
+            form = subdomain1.F
+            for subdomain2 in self.volume_subdomains:
+                jac.append(
+                    dolfinx.fem.form(
+                        ufl.derivative(form, subdomain2.u), entity_maps=entity_maps
+                    )
+                )
+            J.append(jac)
```
Comment (RemDelaporteMathurin) [inline festim/hydrogen_transport_problem.py:1119]: This can be simplified by the PR recently merged by Jorgen

File: festim/hydrogen_transport_problem.py [hunk 13]
```diff
@@ -711,3 +713,506 @@ def post_processing(self):
...
+            forms.append(dolfinx.fem.form(subdomain1.F, entity_maps=entity_maps))
+
+        self.forms = forms
+        self.J = J
```
[the remark this answers is not in the document]
Comment (jorgensd) [reply festim/hydrogen_transport_problem.py:1123]: This should simplify the looping a bit.

File: festim/hydrogen_transport_problem.py [hunk 14]
```diff
@@ -711,3 +713,506 @@ def post_processing(self):
...
+        species=None,
+        reactions=None,
+        temperature=None,
+        sources=None,
+        initial_conditions=None,
+        boundary_conditions=None,
+        settings=None,
+        exports=None,
+        traps=None,
+        interfaces: list[F.Interface] = None,
```
Comment (jorgensd) [inline festim/hydrogen_transport_problem.py:736]: Talk to Remi about proper type-hinting here.

File: festim/hydrogen_transport_problem.py [hunk 15]
```diff
@@ -711,3 +713,506 @@ def post_processing(self):
...
+            v_b = H.subdomain_to_test_function[subdomain_1](b_res)
+            v_t = H.subdomain_to_test_function[subdomain_2](t_res)
+
+            u_b = H.subdomain_to_solution[subdomain_1](b_res)
+            u_t = H.subdomain_to_solution[subdomain_2](t_res)
+
+            K_b = subdomain_1.material.get_solubility_coefficient(
+                self.mesh.mesh, self.temperature_fenics(b_res), H
+            )
+            K_t = subdomain_2.material.get_solubility_coefficient(
```
Comment (jorgensd) [inline festim/hydrogen_transport_problem.py:1073]: Add reference to DG problem and document derivation of this particular way of writing it.

File: festim/mesh/mesh.py [hunk 16]
```diff
@@ -86,10 +88,29 @@ def define_meshtags(self, surface_subdomains, volume_subdomains):
...
+                volume_meshtags.find(domain_1.id),
+                self.vdim,
+                self.fdim,
+            )
+            all_2_facets = dolfinx.mesh.compute_incident_entities(
+                self.mesh.topology,
+                volume_meshtags.find(domain_2.id),
+                self.vdim,
+                self.fdim,
+            )
```
Comment (jhdark) [inline festim/mesh/mesh.py:110]: @jorgensd you mentioned might be better to refer to the domains on either side of an interface by domain_0 and domain_1 right?

File: festim/mesh/mesh_1d.py [hunk 17]
```diff
@@ -72,7 +72,9 @@ def check_borders(self, volume_subdomains):
         ):
             raise ValueError("borders dont match domain borders")
 
-    def define_meshtags(self, surface_subdomains, volume_subdomains):
+    def define_meshtags(self, surface_subdomains, volume_subdomains, interfaces=None):
```
Comment (jhdark) [inline festim/mesh/mesh_1d.py:75]: Might need some better documentation here, but not completely necessary for this PR

File: festim/problem.py
Comment (jhdark) [inline festim/problem.py:1]: Are these changes required for this PR to work? With regards to the progress bar. If not, maybe could leave them for another PR?

File: festim/subdomain/interface.py [hunk 18]
```diff
@@ -0,0 +1,87 @@
...
+            max_cells = np.maximum(mapped_cell_0, mapped_cell_1)
+            self.subdomains[i].parent_to_submesh[integration_data[:, 0]] = max_cells
+            self.subdomains[i].parent_to_submesh[integration_data[:, 2]] = max_cells
+            self.subdomains[i].padded = True
+
+    def compute_mapped_interior_facet_data(self, mesh):
+        """
+        Compute integration data for interface integrals.
+        We define the first domain on an interface as the "+" restriction,
+        meaning that we must sort all integration entities in this order
```
Comment (jhdark) [inline festim/subdomain/interface.py:55]: At the moment these are still quite a black box, could maybe provide some links to documentation and or add some more details to the doc strings

File: festim/hydrogen_transport_problem.py [hunk 19]
```diff
@@ -711,3 +713,506 @@ def post_processing(self):
...
+        n = ufl.FacetNormal(mesh)
+        cr = ufl.Circumradius(mesh)
+
+        entity_maps = {
+            sd.submesh: sd.parent_to_submesh for sd in self.volume_subdomains
+        }
+        for interface in self.interfaces:
+            gamma = interface.penalty_term
+
+            subdomain_1, subdomain_2 = interface.subdomains
```
[the remark this answers is not in the document]
Comment (jhdark) [reply festim/hydrogen_transport_problem.py:1052]: Also could much of this be moved to a method within the `Interface` class?

**Review by jhdark — CHANGES_REQUESTED** [review]
Great progress so far, just a few points from me. I think generally could be important to add more inline comments in places where substantial code blocks will be simplified once integrated into dolfinx. Also needs some more unit tests for specific things such as the new solubility parameters for the material objects, passing the correct meshtags from parent to subdomain. But looking great, doesn't seem far away from merging

**Comment by jhdark** [comment]
Also, you should update the proposed user interface on the PR with the other classes that were introduced

File: festim/hydrogen_transport_problem.py (hunk 11 above)
Comment (RemDelaporteMathurin) [reply festim/hydrogen_transport_problem.py]: done

File: festim/problem.py
Comment (RemDelaporteMathurin) [reply festim/problem.py:1]: No but we're gonna have to suck it up cuase this PR's been open for so long that changes slightly unrelated have been added. Which is why we need to merge this!

File: festim/helpers_discontinuity.py (hunk 1 above)
Comment (jorgensd) [reply festim/helpers_discontinuity.py]: Done

File: festim/hydrogen_transport_problem.py (hunk 4 above)
Comment (jorgensd) [reply festim/hydrogen_transport_problem.py]: Done

File: discontinuity_generic.py (hunk 6 above)
Comment (jorgensd) [reply discontinuity_generic.py]: Done

File: discontinuity_generic.py (hunk 5 above)
Comment (jorgensd) [reply discontinuity_generic.py]: Done

File: discontinuity_generic.py (hunk 8 above)
Comment (jorgensd) [reply discontinuity_generic.py]: Done

File: festim/helpers_discontinuity.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (jorgensd) [reply festim/helpers_discontinuity.py]: Done

File: convergence_rates.py
Comment (RemDelaporteMathurin) [inline convergence_rates.py:1]: We probably don't need to keep this

File: src/festim/exports/maximum_surface.py [hunk 20]
```diff
@@ -23,5 +24,9 @@ def compute(self):
         Computes the maximum value of the field on the defined surface
         subdomain, and appends it to the data list
         """
-        self.value = np.max(self.field.solution.x.array[self.surface.indices])
+        solution = self.field.solution
+        indices = self.surface.locate_boundary_facet_indices(
+            solution.function_space.mesh
+        )
+        self.value = np.max(self.field.solution.x.array[indices])
```
Comment (RemDelaporteMathurin) [inline src/festim/exports/maximum_surface.py:31]: Note for future: this is not safe in parallel

File: src/festim/hydrogen_transport_problem.py [hunk 21]
```diff
@@ -0,0 +1,1264 @@
...
+            exports,
+            traps,
+        )
+        self.interfaces = interfaces or []
+        self.surface_to_volume = surface_to_volume or {}
+        default_petsc_options = {
+            "ksp_type": "preonly",
+            "pc_type": "lu",
+            "pc_factor_mat_solver_type": "mumps",
+        }
```
Comment (RemDelaporteMathurin) [inline src/festim/hydrogen_transport_problem.py:834]: Should live in Settings

File: src/festim/hydrogen_transport_problem.py [hunk 22]
```diff
@@ -0,0 +1,1264 @@
...
+
+            u_b = H.subdomain_to_solution[subdomain_0](res[0])
+            u_t = H.subdomain_to_solution[subdomain_1](res[1])
+
+            K_b = subdomain_0.material.get_solubility_coefficient(
+                self.mesh.mesh, self.temperature_fenics(res[0]), H
+            )
+            K_t = subdomain_1.material.get_solubility_coefficient(
+                self.mesh.mesh, self.temperature_fenics(res[1]), H
+            )
```
Comment (RemDelaporteMathurin) [inline src/festim/hydrogen_transport_problem.py:1116]: Fix bs and ts

File: src/festim/hydrogen_transport_problem.py [hunk 23]
```diff
@@ -0,0 +1,1264 @@
...
+                * (u_b / K_b - u_t / K_t)
+                * v_t
+                * dInterface(interface.id)
+            )
+
+            subdomain_0.F += F_0
+            subdomain_1.F += F_1
+
+        J = []
+        for subdomain1 in self.volume_subdomains:
```
Comment (RemDelaporteMathurin) [inline src/festim/hydrogen_transport_problem.py:1150]:
```suggestion
        # this is the symbolic differentiation of the Jacobian
        for subdomain1 in self.volume_subdomains:
```

File: src/festim/hydrogen_transport_problem.py [hunk 24]
```diff
@@ -0,0 +1,1264 @@
...
+            jac = []
+            for subdomain2 in self.volume_subdomains:
+                jac.append(
+                    ufl.derivative(subdomain1.F, subdomain2.u),
+                )
+            J.append(jac)
+        self.forms = dolfinx.fem.form(
```
Comment (RemDelaporteMathurin) [inline src/festim/hydrogen_transport_problem.py:1157]:
```suggestion
        # compile jacobian (J) and residual (F)
        self.forms = dolfinx.fem.form(
```

File: src/festim/mesh/mesh.py [hunk 25]
```diff
@@ -0,0 +1,127 @@
...
+        if self._mesh is not None:
+            # create cell to facet connectivity
+            self._mesh.topology.create_connectivity(
+                self._mesh.topology.dim, self._mesh.topology.dim - 1
+            )
+
+            # create facet to cell connectivity
+            self._mesh.topology.create_connectivity(
+                self._mesh.topology.dim - 1, self._mesh.topology.dim
+            )
```
Comment (RemDelaporteMathurin) [inline src/festim/mesh/mesh.py:36]: Needs to go in setter
